modelzoo/ESMM/train.py

```diff
@@ -0,0 +1,882 @@
+import argparse
+import collections
+import json
+import math
+import numbers
+import os
+import tensorflow as tf
+from tensorflow.python.ops import partitioned_variables
+from tensorflow.python.client import timeline
+import time
+
+# Set to INFO for tracking training, default is WARN. ERROR for least messages
+tf.logging.set_verbosity(tf.logging.INFO)
+print(f'Using TensorFlow version {tf.__version__}')
+
+USER_COLUMN = [
+    'user_id', 'cms_segid', 'cms_group_id', 'age_level', 'pvalue_level',
+    'shopping_level', 'occupation', 'new_user_class_level'
+]
+ITEM_COLUMN = [
+    'adgroup_id', 'cate_id', 'campaign_id', 'customer', 'brand', 'price'
+]
+COMBO_COLUMN = [
+    'pid', 'tag_category_list', 'tag_brand_list'
+]
+LABEL_COLUMNS = ['clk', 'buy']
+TAG_COLUMN = ['tag_category_list', 'tag_brand_list']
+
+HASH_INPUTS = [
+    'pid', 'adgroup_id', 'cate_id', 'campaign_id', 'customer', 'brand',
+    'user_id', 'cms_segid', 'cms_group_id', 'final_gender_code', 'age_level',
+    'pvalue_level', 'shopping_level', 'occupation', 'new_user_class_level',
+    'tag_category_list', 'tag_brand_list', 'price'
+]
+
+HASH_BUCKET_SIZES = {
+        'pid': 10,
+        'adgroup_id': 100000,
+        'cate_id': 10000,
+        'campaign_id': 100000,
+        'customer': 100000,
+        'brand': 100000,
+        'user_id': 100000,
+        'cms_segid': 100,
+        'cms_group_id': 100,
+        'final_gender_code': 10,
+        'age_level': 10,
+        'pvalue_level': 10,
+        'shopping_level': 10,
+        'occupation': 10,
+        'new_user_class_level': 10,
+        'tag_category_list': 100000,
+        'tag_brand_list': 100000,
+        'price': 50
+        }
+
+defaults = [[0]] * len(LABEL_COLUMNS) + [[' ']] * len(HASH_INPUTS)
+ALL_FEATURE_COLUMNS = HASH_INPUTS
+headers = LABEL_COLUMNS + ALL_FEATURE_COLUMNS
+class ESMM():
+    def __init__(self,
+                 input,
+                 user_column,
+                 item_column,
+                 combo_column,
+                 user_mlp=[256, 128, 96, 64],
+                 item_mlp=[256, 128, 96, 64],
+                 combo_mlp=[128, 96, 64, 32],
+                 cvr_mlp=[128, 96, 64, 32, 16],
+                 ctr_mlp=[128, 96, 64, 32, 16],
+                 batch_size=None,
+                 optimizer_type='adam',
+                 bf16=False,
+                 stock_tf=None,
+                 adaptive_emb=False,
+                 learning_rate=0.1,
+                 l2_scale=1e-6,
+                 input_layer_partitioner=None,
+                 dense_layer_partitioner=None):
+        if not input:
+            raise ValueError('Dataset is not defined.')
+        if not user_column or not item_column or not combo_column:
+            raise ValueError('User column, item column or combo column is not defined.')
+        self.__user_column = user_column
+        self.__item_column = item_column
+        self.__combo_column = combo_column
```

Comment:
Please use a single glide line, to align with other models. I don't really need it to be really private, I just don't really want it to be called externally.